redisvl/utils/vectorize/text/custom.py

```diff
@@ -0,0 +1,293 @@
+
+from pydantic.v1 import PrivateAttr
+
+from redisvl.utils.vectorize.base import BaseVectorizer
+
+
+class CustomTextVectorizer(BaseVectorizer):
+    """The CustomTextVectorizer class wraps user-defined embeding methods to create
+    embeddings for text data.
+
+    This vectorizer is designed to accept a provided callable text vectorizer and
+    provides a class definition to allow for compatibility with RedisVL.
+
+    The vectorizer may support both synchronous and asynchronous operations which
+    allows for batch processing of texts, but at a minimum only syncronous embedding
+    is required to satisfy the 'embed()' method.
+
+    .. code-block:: python
+
+        # Synchronous embedding of a single text
+        vectorizer = CustomTextVectorizer(
+            embed = my_vectorizer.generate_embedding
+        )
+        embedding = vectorizer.embed("Hello, world!")
+
+        # Asynchronous batch embedding of multiple texts
+        embeddings = await vectorizer.aembed_many(
+            ["Hello, world!", "How are you?"],
+            batch_size=2
+        )
+
+    """
+
+    _embed_func: Callable = PrivateAttr()
+    _embed_many_func: Optional[Callable] = PrivateAttr()
+    _aembed_func: Optional[Callable] = PrivateAttr()
+    _aembed_many_func: Optional[Callable] = PrivateAttr()
+
+    def __init__(
+        self,
+        embed: Callable,
+        embed_many: Optional[Callable] = None,
+        aembed: Optional[Callable] = None,
+        aembed_many: Optional[Callable] = None,
+        model: str = "custom_vectorizer",
+    ):
+        """Initialize the Custom vectorizer.
+
+        Args:
+            embed (Callable): a Callable function that accepts a string object and returns a list of floats.
+            embed_many (Optional[Callable)]: a Callable function that accepts a list of string objects and returns a list containing lists of floats. Defaults to None.
+            aembed (Optional[Callable]): an asyncronous Callable function that accepts a string object and returns a lists of floats. Defaults to None.
+            aembed_many (Optional[Callable]):  an asyncronous Callable function that accepts a list of string objects and returns a list containing lists of floats. Defaults to None.
+
+        Raises:
+            ValueError if any of the provided functions accept or return incorrect types.
+            TypeError if any of the provided functions are not Callable objects.
+        """
+
+        self._validate_embed(embed)
+        self._embed_func = embed
+        if embed_many:
+            self._validate_embed_many(embed_many)
+            self._embed_many_func = embed_many
+
+        if aembed:
+            self._validate_aembed(aembed)
+            self._aembed_func = aembed
+        if aembed_many:
+            self._validate_aembed_many(aembed_many)
+            self._aembed_many_func = aembed_many
+
+        super().__init__(model=model, dims=self._set_model_dims())
+
+    def _validate_embed(self, func: Callable):
```

Comment:
Given this is wrapping a pydantic `BaseModel` object, these validators should be implemented using the `validator` operators. (there should be some other examples of this here)

Reply:
Wrapping these with `@validator` decorators calls them before the embed methods are set in `super().__init__`. Pydantic enforces a certain order in initialization so we end up getting trapped in a circular dependency.